### PR TITLE
[EXPERIMENT] Ownership analysis [DO NOT MERGE]

### PR DESCRIPTION
**Pull request by xFrednet**
TL;DR: I want to collect data about ownership in Rust's ecosystem. I modified Clippy to extract things from MIR. This PR is intended for a crater run, but should definitely not be merged.

---

*Roses are red,*
*Violets are blue,*
*Can I borrow,*
*A pattern from you?*

So, what is this project all about? Well, last year I was looking for a topic for my master's thesis. I had the incredible luck that the programming languages group at my university is not only a collection of incredibly cool people but also has some interest in Rust. It just so happens that @amandasystems is part of the group.

We then spent some time looking for a thesis project in the domain of Rust's memory ownership model, and finally settled on the idea to analyze how the memory ownership model is used in practice.

For Polonius, it could be interesting to know how many bodies contain borrows and if the changes from the NLL RFC are actually used, etc.

### Implementation

Disclaimer, this PR contains some of the most convoluted and spaghettified code I've ever written. Do not use this as an example for anything! I'm pretty sure that the insufficiencies of this PR single-handedly increases the compile time by 100%. It's honestly a miracle that the whole thing doesn't ICE on every function.

So at the beginning, I decided to use MIR as that is the IR that the borrow check uses, and it contains all the borrows and moves which might not be explicit in the HIR. While all nice and good it turns out that MIR also gets writ of information which I would have wanted for my analysis.

To be clear, MIR is an excellent tool for what it was designed to do, but its loss of information has been a struggle for this project specifically. 

I would like to collect way more metadata, more complex patterns, and data how named references are used, but the deadline for my thesis is coming up. We decided to limit the scope to named owned variables and see how the analysis goes. If I find some interesting things there is potential to extend this work.

### Run

I decided to implement the analysis on Clippy for the shorter compile times and familiar codebase. This branch disables all other lints by default to safe on some runtime. The crater run with Clippy from master might also expose some ICEs which could be interesting for us. 

The results of this will be aggregated and published as part of my thesis. The time imposed limitations of this implementation probably mean that the results have to be taken with a metric ton of salt, but they should be sufficient to pass my thesis and know if more work in this area would be interesting.

I don't know if there is much more too to say. Let's hope:
* The CI passes
* This run goes smoothly
* Results in some interesting data
* Teaches me how to implement the whole thing better the next time.

---

r? @ghost

### CODE REVIEW

**Comment by xFrednet** [comment]
It's green :D 

Now let's see bors destroy all my hopes and dreams. /s JK love you bors

@bors try

**Comment by Mark-Simulacrum** [comment]
@craterbot run mode=clippy

**Comment by Mark-Simulacrum** [comment]
@craterbot abort

Need to add clippy to try CI.

**Comment by Mark-Simulacrum** [comment]
@craterbot run mode=clippy

**Comment by Mark-Simulacrum** [comment]
@craterbot p=1

**Comment by xFrednet** [comment]
Thank you @Mark-Simulacrum, now it will run before the `beta-1.79-2` test, should be increase the prio of that one to make sure that it runs first or is it not time critical?

**Comment by xFrednet** [comment]
I'm closing this since the experiment is done running. Thank you for kicking off crater!

Now I'll need to take some time to look at the data. :)
